### Big Data Analytics and Visualization 2

### Problem 1 - Starbucks nutrition data

Use the accompanied Starbucks nutrition data (`starbucks-drinks.csv`) to do the following. 
Use separate cells for each part.

1. First, read the data into a pandas dataframe. Show the first few rows.
1. Rename the columns to have no spaces or other [special characters](https://www.owasp.org/index.php/Password_special_characters). 
    - For example, you can remove the units in the column names and change the rest to [snake_case](https://en.wikipedia.org/wiki/Snake_case). 
    - Check the dataframe again to make sure the columns are indeed renamed.
1. What are the mean and median values of calories from all the items?
1. How many items have no less than 250 calories? (Hint: you can use a filter to find out.)
1. Which item(s) have the highest percentage of calories from protein? (Hint: protein has 4 calories per gram)


In [1]:
import pandas as pd
import numpy as np

**Problem 1.1**

In [2]:
data = pd.read_csv("starbucks-drinks.csv")

In [3]:
data.head()

Beverage_category       Beverage      Beverage_prep  Calories  \
0                   Coffee  Brewed Coffee              Short         3   
1                   Coffee  Brewed Coffee               Tall         4   
2                   Coffee  Brewed Coffee             Grande         5   
3                   Coffee  Brewed Coffee              Venti         5   
4  Classic Espresso Drinks    Caffè Latte  Short Nonfat Milk        70   

  Total Fat (g)  Trans Fat (g)  Saturated Fat (g)  Sodium (mg)  \
0           0.1            0.0                0.0            0   
1           0.1            0.0                0.0            0   
2           0.1            0.0                0.0            0   
3           0.1            0.0                0.0            0   
4           0.1            0.1                0.0            5   

   Total Carbohydrates (g)  Cholesterol (mg)  Dietary Fibre (g)  Sugars (g)  \
0                        5                 0                  0           0   
1                       10                 0                  0           0   
2                       10                 0                  0           0   
3                       10                 0                  0           0   
4                       75                10                  0           9   

   Protein (g) Vitamin A (% DV) Vitamin C (% DV) Calcium (% DV) Iron (% DV)  \
0          0.3               0%               0%             0%          0%   
1          0.5               0%               0%             0%          0%   
2          1.0               0%               0%             0%          0%   
3          1.0               0%               0%             2%          0%   
4          6.0              10%               0%            20%          0%   

  Caffeine (mg)  
0           175  
1           260  
2           330  
3           410  
4            75

**Problem 1.2**

In [4]:
data.rename(columns = {'Total Fat (g)':'Total_Fat', 'Trans Fat (g)':'Trans_Fat', 'Saturated Fat (g)':'Total_Fat', 'Sodium (mg)': 'Sodium', 'Total Carbohydrates (g)':'Total_Carbohydrates', 'Cholesterol (mg)':'Cholesterol', 'Dietary Fibre (g)':'Dietary_Fibre', 'Sugars (g)':'Sugars', 'Protein (g)':'Protein', 'Vitamin A (% DV)':'Vitamin_A', 'Vitamin C (% DV)':'Vitamin_C', 'Calcium (% DV)':'Calcium', 'Iron (% DV)':'Iron', 'Caffeine (mg)':'Caffeine'},inplace=True)

In [5]:
data.head()

Beverage_category       Beverage      Beverage_prep  Calories  \
0                   Coffee  Brewed Coffee              Short         3   
1                   Coffee  Brewed Coffee               Tall         4   
2                   Coffee  Brewed Coffee             Grande         5   
3                   Coffee  Brewed Coffee              Venti         5   
4  Classic Espresso Drinks    Caffè Latte  Short Nonfat Milk        70   

   Total_Fat  Trans_Fat  Total_Fat  Sodium  Total_Carbohydrates  Cholesterol  \
0        0.1        0.0        0.0       0                    5            0   
1        0.1        0.0        0.0       0                   10            0   
2        0.1        0.0        0.0       0                   10            0   
3        0.1        0.0        0.0       0                   10            0   
4        0.1        0.1        0.0       5                   75           10   

   Dietary_Fibre  Sugars  Protein Vitamin_A Vitamin_C Calcium Iron Caffeine  
0              0       0      0.3        0%        0%      0%   0%      175  
1              0       0      0.5        0%        0%      0%   0%      260  
2              0       0      1.0        0%        0%      0%   0%      330  
3              0       0      1.0        0%        0%      2%   0%      410  
4              0       9      6.0       10%        0%     20%   0%       75

**Problem 1.3**

In [6]:
data["Calories"].mean()

193.87190082644628

In [7]:
data["Calories"].median()

185.0

**Problem 1.4**

In [8]:
data.query("`Calories`<250").shape

(171, 18)

There are total **171** products  with less than 250 calories

**Problem 1.5**

In [9]:
data["Calories/Protein"] = data["Protein"] * 4

In [10]:
data["Protein %"] = round(data["Calories/Protein"] / data["Calories"] , 3)

In [11]:
data

Beverage_category                                      Beverage  \
0                        Coffee                                 Brewed Coffee   
1                        Coffee                                 Brewed Coffee   
2                        Coffee                                 Brewed Coffee   
3                        Coffee                                 Brewed Coffee   
4       Classic Espresso Drinks                                   Caffè Latte   
..                          ...                                           ...   
237  Frappuccino® Blended Crème  Strawberries & Crème (Without Whipped Cream)   
238  Frappuccino® Blended Crème          Vanilla Bean (Without Whipped Cream)   
239  Frappuccino® Blended Crème          Vanilla Bean (Without Whipped Cream)   
240  Frappuccino® Blended Crème          Vanilla Bean (Without Whipped Cream)   
241  Frappuccino® Blended Crème          Vanilla Bean (Without Whipped Cream)   

          Beverage_prep  Calories  Total_Fat  Trans_Fat  Total_Fat  Sodium  \
0                 Short         3        0.1        0.0        0.0       0   
1                  Tall         4        0.1        0.0        0.0       0   
2                Grande         5        0.1        0.0        0.0       0   
3                 Venti         5        0.1        0.0        0.0       0   
4     Short Nonfat Milk        70        0.1        0.1        0.0       5   
..                  ...       ...        ...        ...        ...     ...   
237             Soymilk       320        3 2        0.4        0.0       0   
238    Tall Nonfat Milk       170        0.1        0.1        0.0       0   
239          Whole Milk       200        3.5        2.0        0.1      10   
240             Soymilk       180        1.5        0.2        0.0       0   
241  Grande Nonfat Milk       240        0.1        0.1        0.0       5   

     Total_Carbohydrates  Cholesterol  Dietary_Fibre  Sugars  Protein  \
0                      5            0              0       0      0.3   
1                     10            0              0       0      0.5   
2                     10            0              0       0      1.0   
3                     10            0              0       0      1.0   
4                     75           10              0       9      6.0   
..                   ...          ...            ...     ...      ...   
237                  250           67              1      64      5.0   
238                  160           39              0      38      4.0   
239                  160           39              0      38      3.0   
240                  160           37              1      35      3.0   
241                  230           56              0      55      5.0   

    Vitamin_A Vitamin_C Calcium Iron Caffeine  Calories/Protein  Protein %  
0          0%        0%      0%   0%      175               1.2      0.400  
1          0%        0%      0%   0%      260               2.0      0.500  
2          0%        0%      0%   0%      330               4.0      0.800  
3          0%        0%      2%   0%      410               4.0      0.800  
4         10%        0%     20%   0%       75              24.0      0.343  
..        ...       ...     ...  ...      ...               ...        ...  
237        6%        8%     20%  10%        0              20.0      0.062  
238        6%        0%     10%   0%        0              16.0      0.094  
239        6%        0%     10%   0%        0              12.0      0.060  
240        4%        0%     10%   6%        0              12.0      0.067  
241        8%        0%     15%   0%        0              20.0      0.083  

[242 rows x 20 columns]

In [12]:
data.sort_values(by="Protein %", ascending=False)

Beverage_category                                    Beverage  \
2                     Coffee                               Brewed Coffee   
3                     Coffee                               Brewed Coffee   
1                     Coffee                               Brewed Coffee   
0                     Coffee                               Brewed Coffee   
41   Classic Espresso Drinks                             Caffè Americano   
..                       ...                                         ...   
166    Shaken Iced Beverages  Shaken Iced Tazo® Tea (With Classic Syrup)   
102         Tazo® Tea Drinks                                   Tazo® Tea   
103         Tazo® Tea Drinks                                   Tazo® Tea   
104         Tazo® Tea Drinks                                   Tazo® Tea   
105         Tazo® Tea Drinks                                   Tazo® Tea   

    Beverage_prep  Calories  Total_Fat  Trans_Fat  Total_Fat  Sodium  \
2          Grande         5        0.1        0.0        0.0       0   
3           Venti         5        0.1        0.0        0.0       0   
1            Tall         4        0.1        0.0        0.0       0   
0           Short         3        0.1        0.0        0.0       0   
41           Tall        10          0        0.0        0.0       0   
..            ...       ...        ...        ...        ...     ...   
166          Tall        60          0        0.0        0.0       0   
102         Short         0          0        0.0        0.0       0   
103          Tall         0          0        0.0        0.0       0   
104        Grande         0          0        0.0        0.0       0   
105         Venti         0          0        0.0        0.0       0   

     Total_Carbohydrates  Cholesterol  Dietary_Fibre  Sugars  Protein  \
2                     10            0              0       0      1.0   
3                     10            0              0       0      1.0   
1                     10            0              0       0      0.5   
0                      5            0              0       0      0.3   
41                    10            2              0       0      1.0   
..                   ...          ...            ...     ...      ...   
166                    0           15              0      15      0.0   
102                    0            0              0       0      0.0   
103                    0            0              0       0      0.0   
104                    0            0              0       0      0.0   
105                    0            0              0       0      0.0   

    Vitamin_A Vitamin_C Calcium Iron Caffeine  Calories/Protein  Protein %  
2          0%        0%      0%   0%      330               4.0        0.8  
3          0%        0%      2%   0%      410               4.0        0.8  
1          0%        0%      0%   0%      260               2.0        0.5  
0          0%        0%      0%   0%      175               1.2        0.4  
41         0%        0%      2%   0%      150               4.0        0.4  
..        ...       ...     ...  ...      ...               ...        ...  
166        0%        0%      0%   6%      170               0.0        0.0  
102        0%        0%      0%   0%   Varies               0.0        NaN  
103        0%        0%      0%   0%   Varies               0.0        NaN  
104        0%        0%      0%   0%   Varies               0.0        NaN  
105        0%        0%      0%   0%   Varies               0.0        NaN  

[242 rows x 20 columns]

From the above we can coclude that the item(s) have the highest percentage of calories from protein are the **Grande, Venti, and Tall type of Brewed coffee.**

### Problem 2 - Marathon time 

Use the accompanied marathon time data (`marathon.csv`) for this problem. 

- The "**split time (s)**" column records the time in seconds when a runner reached the halfway point of the race. 
- The "**final time (s)**" column records a runner's finish time in seconds. 

Do the following. Use separate cells for each part.

1. First, read the data into a pandas DataFrame.
    - Show the first few rows.
    - Show the data types of each column.
    - Show the number of records (i.e., rows) in the data.
    - Delete the 'bib number' column. Then check the first few rows of the dataframe to make sure it's successfully deleted.
    - <font color="gray">Optionally, rename the columns to have no spaces or other <a href=https://www.owasp.org/index.php/Password_special_characters>special characters<a/>.</font>

1. Find the answers to the following questions:
    - How many men (M) are in the data? 
    - How many women (W) are in the data? 
    - How many men in the age group 18-34, inclusive, are in the data? 
    - How many women in the above age group are in the data?

1. The [qualifying time](https://www.baa.org/races/boston-marathon/qualify) of the prestigious Boston Marathon is 3 hours for men and 3 hours 30 minutes for women in the age group of 18-34. What is the percentage of men in this age group who qualified for Boston? Round your answer to 3 decimal places (e.g., 0.456 or 45.6%).
    
1. In the competitive marathon, it's generally better to have a [negative split](https://en.wikipedia.org/wiki/Negative_split) (although very hard to do), which is to run the second half of the race faster than the first half. What is the percentage of the runners in our data who had a negative split? Round your answers to 3 decimal places.
    
1. What is the 5th percentile of the marathon finishing time for all runners (i.e., the time that seperates the fastest 5% of the runners with the remaining runners)? <font color="gray">Optionally, can you convert your answer to hours and minutes format (e.g., 4 hours 30 minutes)?</font>

1. Use the pandas `groupby` method to find the median age and median finish time for (1) all men and (2) all women. 

**Peoblem 2.1**

In [13]:
df = pd.read_csv("marathon.csv")

In [14]:
#First few rows
df.head()

bib number  age gender  split time (s)  final time (s)
0       33677   33      M            3938            7731
1       34433   32      M            3986            7768
2       30615   31      M            4009            7842
3       42492   38      M            3976            8025
4       33481   31      M            3992            8039

In [15]:
#datatype of each column
df.dtypes

bib number         int64
age                int64
gender            object
split time (s)     int64
final time (s)     int64
dtype: object

In [16]:
#Number of records
len(df)

37250

In [17]:
#Deleting the column bib number
df.drop('bib number', inplace=True, axis=1)

In [18]:
df.head()

age gender  split time (s)  final time (s)
0   33      M            3938            7731
1   32      M            3986            7768
2   31      M            4009            7842
3   38      M            3976            8025
4   31      M            3992            8039

In [19]:
#renaming the column headings with no space
df.rename(columns={'split time (s)':'split_time', 'final time (s)':'final_time'}, inplace=True)

In [20]:
df.head()

age gender  split_time  final_time
0   33      M        3938        7731
1   32      M        3986        7768
2   31      M        4009        7842
3   38      M        3976        8025
4   31      M        3992        8039

**Problem 2.2**

In [21]:
df["gender"].value_counts()

gender
M    24665
W    12585
Name: count, dtype: int64

Therefor there are **24665** (M) in the data.

Also there are **12585** (W) in the data.

In [22]:
df1 = df.query("18<=`age`<=34")

In [23]:
df1

age gender  split_time  final_time
0       33      M        3938        7731
1       32      M        3986        7768
2       31      M        4009        7842
4       31      M        3992        8039
5       31      M        3973        8051
...    ...    ...         ...         ...
37226   24      M       12841       31776
37228   24      W       12921       31847
37233   27      W       13069       32261
37244   32      W       15333       34376
37245   18      M       15864       34377

[11164 rows x 4 columns]

In [24]:
df1["gender"].value_counts()

gender
M    6231
W    4933
Name: count, dtype: int64

Therefore there are **5478** men in the age group 18 - 34

Also there are **4536** women in the age group 18 - 34

**Problem 2.3**

In [25]:
#to get the dataframe for only men of 18-34 age group.
dfm= df1[df1["gender"]=="M"]

In [26]:
#to get the dataframe for only women of 18-34 age group.
dfw= df1[df1["gender"]=="W"]

In [27]:
#Now to get the percentage of men qualified for boton,
dfm[dfm["final_time"]<= 10800].value_counts().sum()

394

This gives us that out of 5478 men of the age group 18-34 **394** men are qualified

In [28]:
Total_men = 6231
Qualified_men = 394
print("Percentage of men Qualified = ", round(Qualified_men/Total_men, 3))

Percentage of men Qualified =  0.063


So we get **6.3%** of men qualified

**Problem 2.4**

In [29]:
#Adding new columns for the negative split time, 

In [30]:
df["2nd_half"] = df["final_time"] - df["split_time"]

In [31]:
#STD (Split time difference)
df["STD"] = df["2nd_half"] - df["split_time"]

In [32]:
df

age gender  split_time  final_time  2nd_half   STD
0       33      M        3938        7731      3793  -145
1       32      M        3986        7768      3782  -204
2       31      M        4009        7842      3833  -176
3       38      M        3976        8025      4049    73
4       31      M        3992        8039      4047    55
...    ...    ...         ...         ...       ...   ...
37245   18      M       15864       34377     18513  2649
37246   36      M       16543       34408     17865  1322
37247   51      M       15755       34420     18665  2910
37248   55      W       17886       36040     18154   268
37249   58      W       17989       36068     18079    90

[37250 rows x 6 columns]

In [33]:
df[df["STD"]<0].value_counts().sum()

251

This gives us that out of 37520, total of **251** runners have negative split time

In [34]:
#For the percentage of the same,
Total_runners = 37520
Fast_runners = 251
print("Percentage of runners with negative split time = ", round(Fast_runners/Total_runners,3))

Percentage of runners with negative split time =  0.007


So we get **0.7%** of runners

**Problem 2.5**

In [35]:
df["final_time"].quantile(0.05)

11556.0

From the above we get the finishing time of **11556** seconds. Any record of final time below this comes in the top 5% of the data.

**Problem 2.6**

In [36]:
dfM= df[df["gender"]=="M"]
dfM

age gender  split_time  final_time  2nd_half   STD
0       33      M        3938        7731      3793  -145
1       32      M        3986        7768      3782  -204
2       31      M        4009        7842      3833  -176
3       38      M        3976        8025      4049    73
4       31      M        3992        8039      4047    55
...    ...    ...         ...         ...       ...   ...
37240   73      M       13561       32491     18930  5369
37243   67      M       15587       33375     17788  2201
37245   18      M       15864       34377     18513  2649
37246   36      M       16543       34408     17865  1322
37247   51      M       15755       34420     18665  2910

[24665 rows x 6 columns]

In [37]:
dfW= df[df["gender"]=="W"]
dfW

age gender  split_time  final_time  2nd_half   STD
28      34      W        4355        8653      4298   -57
30      39      W        4393        8737      4344   -49
32      29      W        4389        8750      4361   -28
40      36      W        4377        8893      4516   139
41      27      W        4387        8907      4520   133
...    ...    ...         ...         ...       ...   ...
37241   42      W       12473       32744     20271  7798
37242   64      W       16039       33058     17019   980
37244   32      W       15333       34376     19043  3710
37248   55      W       17886       36040     18154   268
37249   58      W       17989       36068     18079    90

[12585 rows x 6 columns]

In [39]:
#(1) For men;
df.groupby(by="age")["final_time"].median()

age
17    17684.0
18    18833.0
19    17098.5
20    17970.0
21    17796.5
       ...   
81    26141.0
82    24931.0
83    30672.5
84    27173.0
86    30930.0
Name: final_time, Length: 69, dtype: float64

In [39]:
#(2) For women;
dfW.groupby(by="age")["final_time"].median()

age
17    18196.5
18    19585.0
19    18082.0
20    18862.0
21    19395.0
       ...   
76    16121.0
78    19537.0
80    28926.0
82    24931.0
83    31178.0
Name: final_time, Length: 64, dtype: float64

**THANK YOU**